ptychonn/position/registrator.py

```diff
@@ -0,0 +1,919 @@
+            "sift": SIFTRegistrationAlgorithm,
+            "hybrid": HybridRegistrationAlgorithm,
+        }
+        self.algorithm = self.algorithm_dict[self.configs.registration_method](
+            self.configs, **kwargs
+        )
+        self.algorithm.random_seed = self.random_seed
+
+    def has_registratable_features(self, img, std_threshold=0.003, abs_threshold=None):
+        if np.std(img) < std_threshold:
+            return False
+        if abs_threshold and np.abs(img).max() < abs_threshold:
+            return False
+        return True
+
+    def run(self, previous, current):
+        """
+        Run registration and get offset. The returned offset is the supposed *probe position difference* of the
+        current object relative to the previous one, which is opposite to the offset between the images.
+
+        :param previous: object image of the previous scan point.
+        :param current: object image of the current scan point.
+        :return: np.ndarray.
+        """
+        if self.configs.use_baseline_offsets_for_uncertain_pairs:
+            if (
+                not self.has_registratable_features(
+                    previous, std_threshold=0.003, abs_threshold=0.2
+                )
+            ) or (
+                not self.has_registratable_features(
+                    current, std_threshold=0.003, abs_threshold=0.2
+                )
+            ):
+                logging.debug(
+                    "One or both images appear to be empty. Using baseline offset for this pair."
+                )
+                self.algorithm.status = self.algorithm.status_dict["empty"]
+                return np.array([0, 0])
+        offset = self.algorithm.run(previous, current)
+        return offset
+
+    def get_status(self):
+        return self.algorithm.status
+
+    def get_status_code(self, key):
+        return self.algorithm.status_dict[key]
+
+    def update_tol_by_tol_schedule(self, i_point, tol_schedule):
+        tol_schedule = np.array(tol_schedule)
+        transition_points = tol_schedule[:, 0].astype(int)
+        m = i_point == transition_points
+        if np.count_nonzero(m) == 0:
+            return
+        else:
+            i_stage = np.where(m)[0][-1]
+            new_tol = tol_schedule[i_stage, 1]
+            if np.abs(self.algorithm.tol - new_tol) > 1e-8:
+                self.update_tol(new_tol)
+                logging.debug("Updated tol to {}.".format(new_tol))
+
+    def update_tol(self, new_tol):
+        self.configs.nonhybrid_registration_tol = new_tol
+        self.algorithm.tol = new_tol
+
+
+class RegistrationAlgorithm:
+    def __init__(self, configs: RegistrationConfigDict, *args, **kwargs):
+        self.status_dict = {"ok": 0, "questionable": 1, "bad": 2, "empty": 3}
+        self.status = 0
+        self.random_seed = 123
+        self.tol = configs.nonhybrid_registration_tol
+        self.min_roi_stddev = configs.min_roi_stddev
+        self.debug = False
```

Comment:
My knowledge might be incomplete and inaccurate, but apparently `__debug__` is `True` by default and the only way to turn it off is to run the script with `-O` or set the `PYTHONOPTIMIZE` environment variable. This is opposite to what I wanted since I want the debugging code to not run by default.

Reply:
OK. Makes sense.